### DISCUSSION TITLE
login to O365/Azure with Powershell using MFA

### DISCUSSION THREAD

**Discussion by ghost** [pnp/powershell#332] [Q&A]
Not sure if this is the correct group, but hopefully you can guide me!
I'm trying to use a script to login to o365/azure using MFA, but it's not working! I'm posting the script that works without MFA being enabled (sanitized).
Appreciate your help
see attached
[connecto365-Encrypted .txt](https://github.com/pnp/powershell/files/6036419/connecto365-Encrypted.txt)

**Comment by ToddKlindt** [comment]
If you don't tell us what's not working it's tough to help. :) 

Every module is going to handle MFA a little differently, and it matters a lot which MFA provider you're using. Microsoft? Okta? Ping? I also think it's important to recognize that it's inherently incompatible to both use MFA and run scripts unattended. You're either going to have to provide the MFA part, or you're going to have to find a way to authenticate that doesn't require them.

Here is [Microsoft's documentation ](https://docs.microsoft.com/en-us/microsoft-365/enterprise/connect-to-all-microsoft-365-services-in-a-single-windows-powershell-window?view=o365-worldwide)on connecting to their modules. The MFA part is towards the bottom. `Connect-PnPOnline` has a bunch of different authentication options. You can type `help Connect-PnPOnline -examples` to see some of them. You can start with `Connect-PnPOnline -Interactive -LaunchBrowser`. That should prompt you for whatever MFA requirements you have. 
If your MFA provider supports it, and it doesn't violate your security rules, you can use things like certificates to authenticate against Azure AD with `Connect-PnPOnline`.

**Reply by ghost** [reply]
Thank you for responding.  I'm using Microsoft MFA, but sounds like I should create an ID specifically to run scripts etc? 
The script I used to test was deleted since it wasn't working!  using certificates sounds like a good idea instead of MFA.

**Reply by ToddKlindt** [reply]
I can't speak to the other modules, but with the PnP you can register an Azure App and authenticate with that. It does not require an additional user. Check out the [Register-PnPAzureADApp](https://pnp.github.io/powershell/cmdlets/Register-PnPAzureADApp.html) cmdlet. After you register the Azure AD App you can use the certificate, thumbprint, or secret to authenticate.

**Reply by ToddKlindt** [reply]
@Lou629 , [here is a talk](https://www.youtube.com/watch?v=yiBUWbMRw7o) I did about logging into M365 with PowerShell. The bit on [MFA starts here](https://youtu.be/yiBUWbMRw7o?t=1949). That might help you.
